crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,576 @@
+            .times(2)
+            .returning(|input| {
+                let input_cloned = input.clone();
+                async move { Ok(input_cloned
+                    .iter()
+                    .map(|token| (*token, Ok::<_, PriceEstimationError>(1.0)))
+                    .collect::<HashMap<_, _>>()) }.boxed()
+            });
+
+        let config = Configuration {
+            max_concurrent_requests: NonZeroUsize::new(2),
+            max_batch_len: 20,
+            batch_delay: Duration::from_millis(50),
+            result_ready_timeout: Duration::from_millis(500),
+            broadcast_channel_capacity: 50,
+        };
+
+        let buffered = Arc::new(BufferedRequest::with_config(
+            native_price_batch_fetcher,
+            config,
+        ));
+
+        check_batching_many(buffered, tokens_requested).await;
+    }
+
+    #[tokio::test]
+    async fn batching_no_calls() {
+        let mut native_price_batch_fetcher = MockNativePriceBatchFetching::new();
+        native_price_batch_fetcher
+            .expect_fetch_native_prices()
+            // We are testing the native prices are never called
+            .never();
+        let config = Configuration {
+            max_concurrent_requests: NonZeroUsize::new(2),
+            max_batch_len: 20,
+            batch_delay: Duration::from_millis(50),
+            result_ready_timeout: Duration::from_millis(500),
+            broadcast_channel_capacity: 50,
+        };
+
+        let _buffered = Arc::new(BufferedRequest::with_config(
+            native_price_batch_fetcher,
+            config,
+        ));
+
+        sleep(Duration::from_millis(250)).await;
+    }
+
+    #[tokio::test]
+    async fn batching_many_in_multiple_times_successful_estimates() {
+        let tokens_requested = 20;
+        let mut native_price_batch_fetcher = MockNativePriceBatchFetching::new();
+        native_price_batch_fetcher
+            .expect_fetch_native_prices()
+            // We expect this to be requested exactly two times because there are two batches petitions separated by 250 ms
+            .times(2)
+            .returning(|input| {
+                let input_cloned = input.clone();
+                async move { Ok(input_cloned
+                    .iter()
+                    .map(|token| (*token, Ok::<_, PriceEstimationError>(1.0)))
+                    .collect::<HashMap<_, _>>()) }.boxed()
+            });
+
+        let config = Configuration {
+            max_concurrent_requests: NonZeroUsize::new(2),
+            max_batch_len: 20,
+            batch_delay: Duration::from_millis(50),
+            result_ready_timeout: Duration::from_millis(500),
+            broadcast_channel_capacity: 50,
+        };
+
+        let buffered = Arc::new(BufferedRequest::with_config(
+            native_price_batch_fetcher,
+            config,
+        ));
+
+        check_batching_many(buffered.clone(), tokens_requested).await;
+
+        sleep(Duration::from_millis(200)).await;
```

Comment:
FYI often times you can already get reliable tests with 10ms sleeps because there is basically nothing going on in the runtime during a test. Overall we should try to keep tests running quickly because even a bunch of 200ms sleeps can add up over multiple tests and test runs.

Reply:
makes sense, thanks! I set it to `20 ms` because the `batch_delay` is `10 ms` and I wanted it to be the double :ok_hand: